asf_core_data/pipeline/mcs/process/mcs_epc_joining.py

```diff
@@ -0,0 +1,392 @@
+        property characteristics over time. Defaults to True.
+        drop_epc_address (bool, optional): Whether or not to drop addresses
+        from the EPC records. Useful to keep for determining whether
+        matches are sensible. Defaults to True.
+        verbose (bool, optional): Whether or not to print diagnostic information
+        about the matching, e.g. number of matched records. Defaults to True.
+
+    Returns:
+        Dataframe: Merged MCS and EPC records.
+    """
+
+    print("Forming a matching...")
+    matching = form_matching(df1=hps, df2=epcs)
+
+    # First ensure that all matches are above the matching parameter
+    good_matches = matching[
+        (matching["numerics"] == 1) & (matching["address_score"] >= matching_parameter)
+    ].reset_index()
+
+    if all_records:
+        top_matches = (
+            good_matches.groupby("level_0")
+            # Get all level_1 indices of rows in which address_score is maximal
+            .apply(
+                lambda df: df.loc[
+                    df["address_score"] == df["address_score"].max(), "level_1"
+                ]
+            )
+            .droplevel(1)
+            .reset_index()
+            .rename(columns={"index": "level_0"})
+        )
+
+    else:
+        # Get the level_1 index of the first occurrence at which address_score is maximal
+        top_match_indices = good_matches.groupby("level_0")["address_score"].idxmax()
+
+        # Filter the matches df to just the top matches
+        top_matches = good_matches.loc[top_match_indices].drop(
+            columns=["numerics", "address_score"]
+        )
+
+    print("Joining the data...")
+    merged = (
+        hps.reset_index()
+        # Join MCS records to the index-matching df on MCS index
+        .merge(top_matches, how="left", left_on="index", right_on="level_0")
+        # Then join this merged df to EPC records on EPC index
+        .merge(epcs.reset_index(), how="left", left_on="level_1", right_on="index")
+        # Drop any duplicated or unnecessary columns
+        .drop(
+            columns=[
+                "standardised_postcode_x",
+                "standardised_address_x",
+                "numeric_tokens_x",
+                "ADDRESS1",
+                "ADDRESS2",
+                "POSTTOWN",
+                "POSTCODE",
+                "index_y",
+                "postcode_y",
+                "standardised_postcode_y",
+                "numeric_tokens_y",
+                "level_0",
+            ],
+            errors="ignore",
+        )
+    )
+    merged = merged.rename(
+        columns={
+            "index_x": "original_mcs_index",
+            "level_1": "original_epc_index",
+            "postcode_x": "postcode",
+        }
+    )
+
+    if drop_epc_address:
+        merged = merged.drop(columns="standardised_address_y")
+    else:
+        merged = merged.rename(columns={"standardised_address_y": "epc_address"})
+
```

Comment:
Note sure where that would happen, but I'd love to have the UPRN from EPC in the joined mcs_epc file so I can easily match up the MCS installations with the EPC properties. Currently going through the epc_address, which is not so great.

Reply:
I think this is doable! I haven't tested this with the most recent EPC data but this merges all EPC columns by default so UPRN should be included too